File: 0801-0900/0880-Decoded String at Index/0880-Decoded String at Index.py
class Solution:
    def decodeAtIndex(self, S: str, K: int) -> str:
        n = 0
        for i, c in enumerate(S):
            if c.isdigit():
                n *= int(c)
            else:
                n += 1
            if n >= K:
                break
        for j in range(i, -1, -1):
            K %= n
            if K == 0 and S[j].isalpha():
                return S[j]
            if S[j].isdigit():
                n //= int(S[j])
            else:
                n -= 1

File: 0801-0900/0880-Decoded String at Index/0880-Decoded String at Index.cpp
class Solution {
public:
    string decodeAtIndex(string S, int K) {
        int i = 0;
        long long n = 0;
        while (n < K) {
            if (isdigit(S[i])) {
                n *= S[i] - '0';
            }
            else {
                n += 1;
            }
            i += 1;
        }
        
        string result;
        while (i--) {
            K %= n;
            if (K == 0 && isalpha(S[i])) {
                result = string(1, S[i]);
                break;
            }
            if (isdigit(S[i])) {
                n /= S[i] - '0';
            }
            else {
                n -= 1;
            }
        }
        return result;
    }
};
